gst/join/gstjoin.c

```diff
@@ -0,0 +1,1076 @@
+      sel->active_sinkpad ? gst_object_ref (sel->active_sinkpad) : NULL;
+
+  if (sel->active_sinkpad != pad) {
+    gst_join_set_active_pad (sel, pad);
+  }
+
+  active_sinkpad = pad;
+  /* update the segment on the srcpad */
+  if (GST_BUFFER_PTS_IS_VALID (buf)) {
+    GstClockTime start_time = GST_BUFFER_PTS (buf);
+
+    GST_LOG_OBJECT (pad, "received start time %" GST_TIME_FORMAT,
+        GST_TIME_ARGS (start_time));
+    if (GST_BUFFER_DURATION_IS_VALID (buf))
+      GST_LOG_OBJECT (pad, "received end time %" GST_TIME_FORMAT,
+          GST_TIME_ARGS (start_time + GST_BUFFER_DURATION (buf)));
+
+    GST_OBJECT_LOCK (pad);
+    selpad->segment.position = start_time;
+    GST_OBJECT_UNLOCK (pad);
+  }
+  /* Ignore buffers from pads except the selected one */
+  if (pad != active_sinkpad)
+    goto ignore;
+
+  GST_JOIN_UNLOCK (sel);
+
+  if (prev_active_sinkpad != active_sinkpad) {
+    if (prev_active_sinkpad)
+      g_object_notify (G_OBJECT (prev_active_sinkpad), "active");
+    g_object_notify (G_OBJECT (active_sinkpad), "active");
+    g_object_notify (G_OBJECT (sel), "active-pad");
+  }
+
+  /* if we have a pending events, push them now */
+  if (G_UNLIKELY (prev_active_sinkpad != active_sinkpad
+          || selpad->events_pending)) {
+    gst_pad_sticky_events_foreach (GST_PAD_CAST (selpad), forward_sticky_events,
+        sel);
+    selpad->events_pending = FALSE;
+  }
+
+  if (prev_active_sinkpad) {
+    gst_object_unref (prev_active_sinkpad);
+    prev_active_sinkpad = NULL;
+  }
+
+  if (selpad->discont) {
+    buf = gst_buffer_make_writable (buf);
+
+    GST_DEBUG_OBJECT (pad, "Marking discont buffer %p", buf);
+    GST_BUFFER_FLAG_SET (buf, GST_BUFFER_FLAG_DISCONT);
+    selpad->discont = FALSE;
+  }
+
+  /* forward */
+  GST_LOG_OBJECT (pad, "Forwarding buffer %p with timestamp %" GST_TIME_FORMAT,
+      buf, GST_TIME_ARGS (GST_BUFFER_PTS (buf)));
+
+  res = gst_pad_push (sel->srcpad, buf);
+  GST_LOG_OBJECT (pad, "Buffer %p forwarded result=%d", buf, res);
+
+  GST_JOIN_LOCK (sel);
+
+  selpad->pushed = TRUE;
+
+  GST_JOIN_UNLOCK (sel);
+
+done:
+
+  if (prev_active_sinkpad)
+    gst_object_unref (prev_active_sinkpad);
+  prev_active_sinkpad = NULL;
+
+  return res;
+
+  /* dropped buffers */
+ignore:
+  {
+    gboolean active_pad_pushed = GST_JOIN_PAD_CAST (active_sinkpad)->pushed;
+
+    GST_DEBUG_OBJECT (pad, "Pad not active, discard buffer %p", buf);
+    /* when we drop a buffer, we're creating a discont on this pad */
+    selpad->discont = TRUE;
+    GST_JOIN_UNLOCK (sel);
+    gst_buffer_unref (buf);
+
+    /* figure out what to return upstream */
+    GST_OBJECT_LOCK (selpad);
+    if (selpad->always_ok || !active_pad_pushed)
+      res = GST_FLOW_OK;
+    else
+      res = GST_FLOW_NOT_LINKED;
+    GST_OBJECT_UNLOCK (selpad);
+
+    goto done;
+  }
+flushing:
+  {
+    GST_DEBUG_OBJECT (pad, "We are flushing, discard buffer %p", buf);
+    gst_buffer_unref (buf);
+    res = GST_FLOW_FLUSHING;
+    goto done;
+  }
```

Comment:
This goto and label statements are convoluted. Use it only for error handling and it should be linear.
If you want to use such blcok structures, use some static/inline functions

Reply:
In this case, should we stop the pipeline or just warn them?

Reply:
Thank you. Updated.